### DISCUSSION TITLE
Amounts are incorrect in a lot of cases

### DISCUSSION THREAD

**Discussion by daften** [firefly-iii/firefly-iii#7539] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Since recently i've noticed that reconcillation always gives me a notification that the amount is not covered, but the difference is shown as 0, which was strange, but didn't concern me. However now I see that I can't add money to a piggy bank. The account shows as having 500.00 available, but I can't add 500.
I debugged the code, and the problem is that the amount available returned from \FireflyIII\Repositories\PiggyBank\PiggyBankRepository::leftOnAccount gives back 499.999999999999 instead of 500. I also see now in the database that the transactions for e.g. transfers are often incorrect. See the screenshot in attachment of the transactions table. You'll see that two transactions for a transfer often diverge with 0.000....001, which seems meaningless, but isn't. This also makes me think the bug is in inserting data somewhere. I couldn't trace that yet.

### Debug information

Debug information generated at 2023-05-22 00:00:00 UTC for Firefly III version **6.0.10**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.10 |
| Firefly III API | 2.0.1 |
| DB version | 19 (exp. 19) |
| Docker | No |
| PHP | `8.2.3` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ |  |
| Browser TZ | Europe/Brussels |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 104857600 (100 MB) |
| Interface | fpm-fcgi |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2023-05-01 00:00:00 |
| Session end | 2023-05-31 23:59:59 |
| Session first | 2019-12-25 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.4 Safari/605.1.15 |

**Comment by daften** [comment]
Looking further, I think the error has to be in \FireflyIII\Support\Steam::negative, I honestly don't see what can be done to fix it, since you want to keep using floating-point arithmetic, maybe it was introduced by changes in PHP itself in bc_math and a change that was introduced there? The problem i see is that it's really hard to fix, but also is a problem that will lead to continuously more incorrect data.

Update: I see also a transaction that was imported through the data importer with a CSV file. The amount in the CSV file is 9.10 however in the database BOTH transactions contain 9.099999999999 as an amount.

Update 2: To help trace this, I checked which transactions have an amount or foreing_amount ending in 999. The first one is 07/12/2022, then for months none and on 13/03/2023 they start again and occur quite frequently. I have auto-updates enabled for my docker containers through watchtower, so normally my firefly instance is updated within 1.5 days. Looking at the dates, it seems to more or less coincide with the switch to version 6, which is the only one that might have been delayed maybe.

**Comment by daften** [comment]
I ran this query on a test instance, and it seemed to correct data. But I'm unsure of it. I think all of my transactions are limited to 2 decimal points, but that might not be the case for everybody:

```
UPDATE transactions SET amount = ROUND(amount, 2), foreign_amount = ROUND(foreign_amount, 2)
```

**Comment by JC5** [comment, marked as answer]
There was a bug that caused this, which has since been fixed. Make a backup and run this command:

```
php artisan firefly-iii:force-decimal-size
```

**Comment by daften** [comment]
That fixed it indeed. Apologies, I must have missed this in the release notes. Thanks for the quick reply!!

**Reply by JC5** [reply]
It wasn't in the release notes, the command is kind of experimental. Hence the warning! I will add it in 6.0.11 :)

**Comment by daften** [comment]
Hi,

FYI: I just had more transactions that were wrong. Running the command fixes it, but the bug seems to still be there. I noticed it when performing reconciliation, it was offering to add a corrective transaction for 0.00 again.

**Reply by JC5** [reply]
Thanks for letting me know. I haven't spotted it myself. How do you create transactions?

**Reply by daften** [reply]
- Cash and CC transactions => manually through firefly interface, but those don't have issues afaik.
- CA transactions => through a csv export from my bank and  the data importer. Those are the ones where issues can occur. I already checked CSV exports used and they are fine, so it's somewhere in the import process.

Does that help? :) Sorry for the late reply, some personal stuff popped up over the  weekend.

**Reply by JC5** [reply]
Not a problem, this narrows it down. Traditionally, API transactions were the issue when amounts were directly submitted as floats, ie:

```
"amount": 12.34
```

As opposed to

```
"amount": "12.34"
```

I'll see what I can do re: the data importer. As far as I know, it only submits strings, but you never know. If you happen to be able to collect logs from the data importer, I'm curious to learn what the submission was like for a problematic transaction.

**Comment by daften** [comment]
I wanted to report back that I haven't had it since I've reported the last update here. I've paid attention to all my reconciliations, since that were I notice it, and everything seems to be ok.
